diff --git a/nanosim/src/simulator.py b/nanosim/src/simulator.py
index 1417741..4555e97 100755
--- a/nanosim/src/simulator.py
+++ b/nanosim/src/simulator.py
@@ -379,6 +379,8 @@ def simulation(ref, out, dna_type, per, kmer_bias, max_l, min_l, median_l=None,
 
             # Mutate read
             new_read = case_convert(new_read)
+            if len(new_read) < ref:
+                continue
             read_mutated = mutate_read(new_read, new_read_name, out_error, error_dict, kmer_bias)
 
             # Reverse complement half of the reads
